src/mesh.cpp

```diff
@@ -1852,6 +1963,225 @@ UnstructuredMesh::write(std::string base_filename) const {
+  if (!variable_map_.count(value_name)) {
+    auto& eqn_sys = equation_systems_->get_system(eq_system_name_);
+    auto var_num = eqn_sys.add_variable(value_name, libMesh::CONSTANT, libMesh::MONOMIAL);
+    variable_map_[value_name] = var_num;
+  }
+
+  std::string std_dev_name = var_name + "_std_dev";
+  // check if this is a new variable
+  if (!variable_map_.count(std_dev_name)) {
+    auto& eqn_sys = equation_systems_->get_system(eq_system_name_);
+    auto var_num = eqn_sys.add_variable(std_dev_name, libMesh::CONSTANT, libMesh::MONOMIAL);
+    variable_map_[std_dev_name] = var_num;
+  }
+}
+
+void
+LibMesh::remove_score(const std::string& var_name)
+{
+  return;
+}
+
+
+void
+LibMesh::set_score_data(const std::string& var_name,
+                        std::vector<double> values,
+                        std::vector<double> std_dev)
+{
+  auto& eqn_sys = equation_systems_->get_system(eq_system_name_);
+
+  if (!eqn_sys.is_initialized()) { equation_systems_->init(); }
+
+  const libMesh::DofMap& dof_map = eqn_sys.get_dof_map();
+
+  // look up the value variable
+  std::string value_name = var_name + "_mean";
+  unsigned int value_num = variable_map_.at(value_name);
+  // look up the std dev variable
+  std::string std_dev_name = var_name + "_std_dev";
+  unsigned int std_dev_num = variable_map_.at(std_dev_name);
+
+  for (int bin = 0; bin < this->n_bins(); bin++) {
+    auto e = m_->elem_ptr(bin);
+
+    // set value
+    std::vector<libMesh::dof_id_type> value_dof_indices;
+    dof_map.dof_indices(e, value_dof_indices, value_num);
+    Ensures(value_dof_indices.size() == 1);
+    eqn_sys.solution->set(value_dof_indices[0], values[bin]);
+
+    // set std dev
+    std::vector<libMesh::dof_id_type> std_dev_dof_indices;
+    dof_map.dof_indices(e, std_dev_dof_indices, std_dev_num);
+    Ensures(std_dev_dof_indices.size() == 1);
+    eqn_sys.solution->set(std_dev_dof_indices[0], std_dev[bin]);
+
+  }
+}
+
+void LibMesh::write(std::string filename) const
+{
+  write_message(fmt::format("Writing file: {}.e for unstructured mesh {}", filename, this->id_));
+  libMesh::ExodusII_IO exo(*m_);
+  std::set<std::string> systems_out = {eq_system_name_};
+  exo.write_discontinuous_exodusII(filename + ".e", *equation_systems_, &systems_out);
+}
+
+void
+LibMesh::bins_crossed(const Particle& p,
+                      std::vector<int>& bins,
+                      std::vector<double>& lengths) const
+{
+  // TODO: Implement triangle crossings here
+  throw std::runtime_error{"Tracklength tallies on libMesh instances are not implemented."};
+}
+
+int
+LibMesh::get_bin(Position r) const
+{
+  // look-up a tet using the point locator
+  libMesh::Point p(r.x, r.y, r.z);
+
+  // quick rejection check
+  if (!bbox_.contains_point(p)) { return -1; }
+
+  #ifdef _OPENMP
+  int thread_num = omp_get_thread_num();
+  #else
+  int thread_num = 0;
+  #endif
+
+  const auto& point_locator = PL_.at(thread_num);
+
+  auto e = (*point_locator)(p);
```

Comment:
`e` is a libMesh element pointer. I'll update the name to `elem_ptr` for clarity. It's the only type that can be returned from the `libmesh::PointLocator`. Based on the `libMesh` interface only returning pointers or references and never copies of these objects I think it's probably best to avoid copying this object.

Reply:
My point was that if you do `auto e`, you _will_ get a copy, although if it's returning a pointer then you're effectively getting a copy of the pointer. `auto&` would probably be safer